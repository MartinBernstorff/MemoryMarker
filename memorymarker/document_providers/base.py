from dataclasses import dataclass
from typing import TYPE_CHECKING, Protocol, Sequence

if TYPE_CHECKING:
    import datetime as dt
    from pathlib import Path

    from iterpy.iter import Iter

    from memorymarker.question_generator.reasoned_highlight import Highlights

    from .omnivore import OmnivoreDocument


@dataclass(frozen=True)
class OrphanHighlight:
    """Highlight without a source document"""

    highlight: str
    uri: str
    title: str


class DocumentProvider(Protocol):
    def get_documents(self) -> "Iter[OmnivoreDocument]":
        ...


class HighlightManager(Protocol):
    timestamp_file: "Path"
    source: DocumentProvider

    def get_highlights_since_update(
        self, date: "dt.datetime"
    ) -> Sequence["Highlights"]:
        ...
