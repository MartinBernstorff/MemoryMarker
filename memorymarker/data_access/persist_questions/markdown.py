from dataclasses import dataclass
from pathlib import Path
from typing import Mapping, Sequence

from memorymarker.domain.highlights_to_questions import QAPrompt

SAVE_DIR = Path("card_cache")

import re


def clean_filename(filename: str) -> str:
    # This will replace not allowed symbols with an underscore.
    return re.sub(r"[^A-Za-z]", "_", filename)


def q_to_markdown(prompt: QAPrompt) -> str:
    return f"Q. {prompt.question}\nA. {prompt.answer}\n\n"


def write_md(contents: str, file_title: str, save_dir: Path = SAVE_DIR) -> None:
    """Write markdown to file. Append if exists"""
    if not save_dir.exists():
        save_dir.mkdir()
    filename = f"{clean_filename(file_title)}.md"
    with Path.open(SAVE_DIR / filename, "a") as f:
        f.write(contents)


def write_qa_prompt_to_md(prompt: QAPrompt, save_dir: Path = SAVE_DIR) -> None:
    """Write markdown to file. Append if exists"""
    contents = q_to_markdown(prompt)
    write_md(contents=contents, file_title=prompt.title, save_dir=save_dir)
