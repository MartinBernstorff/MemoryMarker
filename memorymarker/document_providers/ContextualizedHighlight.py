import datetime as dt  # noqa: TCH003

from pydantic import BaseModel


class ContextualizedHighlight(BaseModel):
    source_doc_title: str
    source_doc_uri: str

    prefix: str | None
    highlighted_text: str
    suffix: str | None

    source_highlight_uri: str | None = None
    updated_at: dt.datetime

    @property
    def context(self) -> str:
        context = ""
        context += self.prefix or ""
        context += self.highlighted_text
        context += self.suffix or ""

        return context
