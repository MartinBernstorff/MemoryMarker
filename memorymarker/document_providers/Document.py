from typing import Any, Mapping, Sequence

from iterpy.iter import Iter
from pydantic import BaseModel

from .ContextualizedHighlight import ContextualizedHighlight


class Document(BaseModel):
    title: str
    uri: str
    slug: str
    highlights: Sequence[Mapping[str, Any]]

    def _parse_highlight(self, highlight: Mapping[str, str]) -> ContextualizedHighlight:
        return ContextualizedHighlight(
            source_doc_title=self.title,
            source_doc_uri=self.uri,
            highlighted_text=highlight["quote"],
            prefix=highlight["prefix"],
            suffix=highlight["suffix"],
            updated_at=highlight["updatedAt"],  # type: ignore
            source_highlight_uri=f"https://omnivore.app/me/{self.slug}#{highlight["id"]}",
        )

    def get_highlights(self) -> Iter[ContextualizedHighlight]:
        return Iter(self.highlights).map(self._parse_highlight)
