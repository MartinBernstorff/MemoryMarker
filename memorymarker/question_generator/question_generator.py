import asyncio
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING, Literal, Sequence

import instructor
from openai import AsyncOpenAI
from openai.types.chat.chat_completion_user_message_param import (
    ChatCompletionUserMessageParam,
)

from memorymarker.question_generator.highlight_to_question import HighlightToQuestion

if TYPE_CHECKING:
    from openai.types.chat.chat_completion_message_param import (
        ChatCompletionMessageParam,
    )

    from memorymarker.document_providers.ContextualizedHighlight import (
        ContextualizedHighlight,
    )
    from memorymarker.question_generator.qa_prompt import QAPrompt

PROMPT_DIR = Path(__file__).parent.parent / "prompts"
assert PROMPT_DIR.exists(), f"Prompts directory does not exist at {PROMPT_DIR}"
SYSTEM_PROMPT = (PROMPT_DIR / "martin_prompt.txt").read_text()


from iterpy.iter import Iter

from memorymarker.question_generator.qa_prompt import QAPromptResponseModel

OPENAI_MODELS = Literal[
    "gpt-4-0125-preview",
    "gpt-4-turbo-preview",
    "gpt-4-1106-preview",
    "gpt-4-vision-preview",
    "gpt-4",
    "gpt-4-0314",
    "gpt-4-0613",
    "gpt-4-32k",
    "gpt-4-32k-0314",
    "gpt-4-32k-0613",
    "gpt-3.5-turbo",
    "gpt-3.5-turbo-16k",
    "gpt-3.5-turbo-0301",
    "gpt-3.5-turbo-0613",
    "gpt-3.5-turbo-1106",
    "gpt-3.5-turbo-0125",
    "gpt-3.5-turbo-16k-0613",
]


@dataclass
class BaselinePipeline(HighlightToQuestion):
    openai_api_key: str
    model: OPENAI_MODELS
    prompt: str = SYSTEM_PROMPT

    def __post_init__(self):
        self.client = instructor.patch(  # type: ignore # Incorrectly typed from Instructor library
            AsyncOpenAI(api_key=self.openai_api_key)
        )

    def _build_message(
        self, highlight: "ContextualizedHighlight"
    ) -> "ChatCompletionMessageParam":
        return ChatCompletionUserMessageParam(
            role="user",
            name="Martin",
            content=f"""{self.prompt}

<target>{highlight.highlighted_text}</target><context>{highlight.context}</context>""",
        )

    async def _highlight_to_question(
        self, highlight: "ContextualizedHighlight"
    ) -> "QAPromptResponseModel":
        return await self.client.chat.completions.create(
            model=self.model,
            messages=[self._build_message(highlight=highlight)],
            response_model=QAPromptResponseModel,
            temperature=0.0,
        )  # type: ignore

    async def _gather(
        self, highlights: Sequence["ContextualizedHighlight"]
    ) -> Sequence["QAPromptResponseModel"]:
        questions = [self._highlight_to_question(highlight) for highlight in highlights]
        response = await asyncio.gather(*questions)
        return response

    def __call__(self, highlights: "Iter[ContextualizedHighlight]") -> "Iter[QAPrompt]":
        response: Sequence[QAPromptResponseModel] = asyncio.run(
            self._gather(highlights)  # type: ignore
        )

        hydrated_responses = (
            Iter(response)
            .zip(highlights.to_list())  # type: ignore
            .map(
                lambda qa_context_pair: qa_context_pair[0].to_qaprompt(  # type: ignore
                    qa_context_pair[1]  # type: ignore
                )
            )
        )
        return hydrated_responses
