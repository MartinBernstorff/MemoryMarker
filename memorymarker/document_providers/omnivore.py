import os
from dataclasses import dataclass
from typing import Mapping

from iterpy._iter import Iter
from omnivoreql import OmnivoreQL

from memorymarker.document_providers.Document import Document

from .base import DocumentProvider


@dataclass
class Omnivore(DocumentProvider):
    def __post_init__(self):
        omnivore_api_key = os.getenv("OMNIVORE_API_KEY")
        if not omnivore_api_key:
            raise ValueError("OMNIVORE_API_KEY environment variable not set")
        self.client = OmnivoreQL(omnivore_api_key)

    def _parse_doc(self, document: Mapping[str, str]) -> Document:
        return Document(
            title=document["title"],
            uri=document["url"],
            highlights=document["highlights"],  # type: ignore
            slug=document["slug"],
        )

    def get_documents(self) -> Iter[Document]:
        documents = (
            Iter(self.client.get_articles(limit=1000)["search"]["edges"])
            .map(lambda a: a["node"])
            .map(self._parse_doc)
            .flatten()
        )
        return documents
