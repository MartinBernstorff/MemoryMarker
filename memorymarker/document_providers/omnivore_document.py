from typing import Any, Mapping, Sequence

from iterpy.iter import Iter
from pydantic import BaseModel

from memorymarker.question_generator.reasoned_highlight import (
    ReasonedHighlight,
    SourceDocument,
)


def empty_string_if_none(value: str | None) -> str:
    return value or ""


class OmnivoreDocument(BaseModel):
    title: str
    uri: str
    slug: str
    highlights: Sequence[Mapping[str, Any]]

    def _parse_highlight(
        self, highlight: Mapping[str, str]
    ) -> ReasonedHighlight | None:
        if "quote" not in highlight or highlight["quote"] is None:  # type: ignore
            return None

        return ReasonedHighlight(
            source_document=SourceDocument(
                title=self.title,
                uri=f"https://omnivore.app/me/{self.slug}#{highlight["id"]}",
            ),
            pipeline_name="",
            reasoning_prompt="",
            reasoning="",
            qa_string="",
            question_answer_pairs=[],
            highlighted_text=highlight["quote"],
            prefix=empty_string_if_none(highlight["prefix"]),
            suffix=empty_string_if_none(highlight["suffix"]),
            updated_at=highlight["updatedAt"],  # type: ignore # Will be recast on init.
        )

    def get_highlights(self) -> Iter[ReasonedHighlight]:
        highlights = Iter(self.highlights).map(self._parse_highlight)
        return highlights.filter(lambda _: _ is not None)  # type: ignore
