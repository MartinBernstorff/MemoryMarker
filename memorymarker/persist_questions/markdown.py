import logging
import re
from typing import TYPE_CHECKING, Sequence

if TYPE_CHECKING:
    from pathlib import Path

    from memorymarker.question_generator.reasoned_highlight import ReasonedHighlight


def clean_filename(filename: str) -> str:
    # This will replace not allowed symbols with an underscore.
    without_illegal = re.sub(r"[^A-Za-z\s]", " ", filename)
    without_duplicate_spaces = re.sub(r"\s{2,}", ", ", without_illegal)
    return without_duplicate_spaces


def highlight_group_to_file(
    output_dir: "Path", group: tuple[str, Sequence["ReasonedHighlight"]]
) -> None:
    save_path = output_dir / clean_filename(group[0])

    with save_path.open(mode="a") as f:
        for highlight in group[1]:
            f.write(highlight.to_markdown())
        f.write("\n---\n")

    logging.info(f"Wrote {group[0]}")
