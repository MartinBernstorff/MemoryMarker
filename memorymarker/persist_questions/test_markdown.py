import datetime as dt

import pytest

import memorymarker.persist_questions.markdown as markdown
from memorymarker.question_generator.qa_prompt import QAPrompt

from ..document_providers.contextualized_highlight import ContextualizedHighlight


class FakeHydratedHighlight(ContextualizedHighlight):
    source_doc_title: str = "The Hitchhiker's Guide to the Galaxy"
    source_doc_uri: str = (
        "https://en.wikipedia.org/wiki/The_Hitchhiker%27s_Guide_to_the_Galaxy"
    )
    source_highlight_uri: str | None = "https://en.wikipedia.org/wiki/The_Hitchhiker%27s_Guide_to_the_Galaxy#meaning_of_life"
    updated_at: dt.datetime = dt.datetime.now()

    prefix: str | None = ""
    highlighted_text: str = "42"
    suffix: str | None = ""


@pytest.fixture(scope="module")
def question() -> QAPrompt:
    return QAPrompt(
        question="What is the meaning of life?",
        answer="42",
        title="The Hitchhiker's Guide to the Galaxy",
        hydrated_highlight=FakeHydratedHighlight(),
    )


def test_single_q_to_markdown(question: QAPrompt, snapshot) -> None:  # noqa: ANN001
    input_md = markdown.q_to_markdown(question)
    assert snapshot == input_md


def test_clean_filename():
    filename = "Wikipedia Â· Powerhouse of the cell - Testing"
    result = markdown.clean_filename(filename)
    assert result == "Wikipedia___Powerhouse_of_the_cell___Testing"
