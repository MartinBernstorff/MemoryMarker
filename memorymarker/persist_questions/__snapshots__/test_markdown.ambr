# serializer version: 1
# name: test_single_q_to_markdown
  '''
  Q) What is the meaning of life?
  
  > [!NOTE]- Highlight
  >  ==42== 
  > [Link](https://en.wikipedia.org/wiki/The_Hitchhiker%27s_Guide_to_the_Galaxy#meaning_of_life)
  
  
  '''
# ---
